Optimize universe nodes and handle r-values in Python backend.

notoraptor opened the pull request:
Hi @abergeron @breuleux This is a PR to optimize universe nodes in Python backend.

I also fix some bugs related to compilation of literal strings, which were previously handled as symbol names (`string` instead of `"string"`).

notoraptor committed:
- Convert node inputs to r-values, to handle litteral strings correctly.
- Add a test for print and check if litteral string passed to print is correctly compiled.

notoraptor committed:
Convert inline return to r-value, to handle litteral strings correctly.
Add a test to check if a litteral string is correctly compiled as a return statement.

notoraptor committed:
Move test_string() into test_code_format to check compiled code.

notoraptor committed:
Convert default formatter inputs to r-values.
Add a test for operator add with strings.

notoraptor committed:
Convert node inputs to r-values for special implementations.
Add a test to check that literal strings passed to special implementations (here universe_setitem) are correctly compiled.

notoraptor committed:
Implement `typeof` directly using `type`.

notoraptor committed:
When parsing myia graph, skip known parent nodes and known local arrows, instead of skipping known nodes.
This allows to not miss any arrow, while still preventing from handling a parent node in a closure.

notoraptor committed:
Add a Python backend optimizer to optimize directed graphs bfore code generation.

Current optimizer optimizes nodes:
- typeof
- make_handle
- universe_setitem
- universe_getitem

breuleux commented on myia/compile/backends/python/code_generator.py:
```diff
@@ -190,22 +215,22 @@ def _register_global(self, name, value):
     def _node_to_expr(self, node):
         """Convert an apply node to an expr."""
-        fn = node.fn
+        node = self.replace.get(node, node)
+        fn = self.replace.get(node.fn, node.fn)
+        inputs = [self.replace.get(inp, inp) for inp in node.inputs]
```
Is it possible that a node's replacement might itself have a replacement?

notoraptor replied:
Hi! Currently no, it's just one replacement by a new node, no recursion.

I don't think I will have to handle recursive recursion in the future, unless very complex optimization to do in backend.

breuleux commented on myia/compile/backends/python/code_generator.py:
```diff
@@ -224,14 +249,15 @@ def _node_to_line(self, node):
-            if node in self.inline_nodes
-            else self.lbl(node)
-        )
+        node = self.replace.get(node, node)
```
Is this necessary? I assume that we already got the node's replacement in the caller to this function.

breuleux commented on myia/compile/backends/python/code_generator.py:
```diff
@@ -224,14 +249,15 @@ def _node_to_line(self, node):
 
     def rvalue(self, node):
         """Inline constant strings and get name for other nodes."""
+        node = self.replace.get(node, node)
```
Same question as for label.

breuleux commented on myia/compile/backends/python/optimizer.py:
```diff
@@ -0,0 +1,105 @@
+        """Optimize `universe_getitem` nodes for given directed graph."""
+        nodes_universe_getitem = []
+        for element in dg.value_to_node:
+            if isinstance(element, Apply):
+                fn = element.fn
+                if fn.is_constant() and fn.value == UNIVERSE_GETITEM:
+                    nodes_universe_getitem.append(element)
```
The utility function `element.is_apply(UNIVERSE_GETITEM)` should take care of both conditions here.

notoraptor replied:
Ok !

breuleux commented on myia/compile/backends/python/optimizer.py:
```diff
@@ -0,0 +1,105 @@
+    def _optimize_universe_setitem(self, dg: DirectedGraph):
+        """Optimize `universe_setitem` nodes for given directed graph."""
+        # Dictionary mapping typename to myia node to assignation object
+        nodes = {TYPEOF: set(), MAKE_HANDLE: set(), UNIVERSE_SETITEM: set()}
+        for element in dg.value_to_node:
+            if isinstance(element, Apply):
+                fn = element.fn
+                if fn.is_constant() and fn.value in nodes:
+                    assert element not in nodes[fn.value]
+                    nodes[fn.value].add(element)
```
I think the design for this could be made somewhat more generic, by e.g. having a function like `self.collect_nodes({TYPEOF, MAKE_HANDLE, UNIVERSE_SETITEM})` return the nodes dict you have here. That function could be reused for `_optimize_universe_getitem`, with a different argument.

notoraptor replied:
Ok !

notoraptor committed:
Optimizer: add a utility function `_collect_apply_nodes()` to collect apply nodes with specified functions.

notoraptor committed:
Reorganize code_generator code.
Apply node replacement into:
- `directed_graph_to_node()`, before passing node to other methods
- `_node_to_expr()` for apply fn and inputs, before using them
- `label()` as this method may be called to a node not-yet replaced
Rename `rvalue()` to private `_rvalue()` which is always called to an already-replaced node
Add a private `_label()`, used by `label()`, and always called to an already-replaced node.

notoraptor commented:
Hi @breuleux , PR updated !

I re-organized the `code_generator` code for a better follow-up of function usages.
- `rvalue()` indeeds always receives an already-replaced node. It is now a private method `_rvalue()`.
- `label()` may be called from outside code generator (in Python backend to get main graph name), so it still replaces node if necessary (although I currently replace only nodes, not graphs). Anyway, I added a method `_label()`, for specific labeling of already-replaced nodes. `label()` uses `_label()`.
